src/core/ext/filters/client_channel/resolver/xds/xds_resolver.cc

```diff
@@ -741,6 +741,10 @@ void XdsResolver::OnListenerUpdate(XdsApi::LdsUpdate listener) {
   if (route_config_name_.empty()) {
     GPR_ASSERT(current_listener_.rds_update.has_value());
     OnRouteConfigUpdate(std::move(*current_listener_.rds_update));
+  } else {
+    // HCM may contain newer filter config. We need to propagate the update as
+    // config selector to the channel
+    GenerateResult();
```

Comment:
Consider the case where we get an LDS update that changes both the filter list and the RDS resource name.  In that case, we do not want to generate a result here, because we want to swap in both changes at the same time -- i.e., we don't want to swap in the new filter list until we can also apply the new RouteConfiguration.

To fix this, I suggest calling `current_virtual_host_.Clear()` on line 735 above.  That way, `GenerateResult()` will be a no-op in the case where we've switched to a different RDS resource name.

Reply:
Emm... I couldn't find the `Clear()` function mentioned here. I try to submit the change to see if I missed any abstraction layer but the build has failed. And the VirtualHost struct doesn't have watchers to be cleared.

```
struct VirtualHost {
      std::vector<std::string> domains;
      std::vector<Route> routes;
      TypedPerFilterConfig typed_per_filter_config;
}
```

Should we add a clear function to clean all fields here? @markdroth

Reply:
Ah, sorry, not sure what I was thinking there.  Instead, you can do `current_virtual_host_.routes.clear()`.  We don't really need to clear the other fields, because `GenerateResult()` only checks this one.